### PR TITLE
ASE: Added a libopae-c-ase.so to make ASE application backward compat…

### PR DESCRIPTION
**Pull request by lzhan55**
…ible

### CODE REVIEW

File: libopae/init_ase.c [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+#define __USE_GNU
+#endif // __USE_GNU
+#include <pthread.h>
+
+#define PATH_MAX 1024
+
+__attribute__((constructor)) STATIC void opae_ase_init(void)
+{
+	char cfg_path[PATH_MAX];
+	char *opae_path = getenv("OPAE_PLATFORM_ROOT");
```
Comment (michael-adler) [inline libopae/init_ase.c]: Shouldn't this be searching the OPAE SDK installation directory (where ASE is stored), not the release directory?

Comment (lzhan55) [reply libopae/init_ase.c]: @michael-adler Do we have an environment variable for the OPAE SDK installation directory?

Comment (michael-adler) [reply libopae/init_ase.c]: No, but it is known at build time since it is all inside OPAE.

You can see an example in getASESrcPath() in https://github.com/OPAE/opae-sdk/blob/4a7de2e54c0dd308c00d5386a2736eebd3c16b64/ase/scripts/afu_sim_setup#L46. There has to be a special CMake rule to do the substitution, so a bit messy, but definitely possible.

File: CMakeLists.txt [hunk 2]
```diff
@@ -29,7 +29,7 @@ project(opae)
 
 ############################################################################
 ## Get git hash info, if available #########################################
-############################################################################
+############################################################################env
```
Comment (tswhison) [inline CMakeLists.txt]: Remove env at end of line.

File: CMakeLists.txt [hunk 3]
```diff
@@ -113,6 +113,10 @@ install(DIRECTORY common/include/opae
 install(DIRECTORY common/include/safe_string
   DESTINATION include
   COMPONENT safestrheaders)
+  
+install(FILES libopae/plugins/ase/opae_ase.cfg     
+  DESTINATION include
```
Comment (tswhison) [inline CMakeLists.txt]: Why in the includes folder?  We won't search there.

File: libopae/CMakeLists.txt [hunk 4]
```diff
@@ -24,7 +24,7 @@
 ## ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
 ## POSSIBILITY OF SUCH DAMAGE.
 
-project(libopae-c)
+project(libopae-c)
```
Comment (tswhison) [inline libopae/CMakeLists.txt]: Remove space at end.

File: libopae/init_ase.c [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+#include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#ifndef __USE_GNU
+#define __USE_GNU
+#endif // __USE_GNU
+#include <pthread.h>
+
+#define PATH_MAX 1024
```
Comment (tswhison) [inline libopae/init_ase.c]: Remove this and include limits.h instead.  It's defined there as 4096.

File: libopae/init_ase.c [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+{
+	char cfg_path[PATH_MAX];
+	char *opae_path = getenv("OPAE_PLATFORM_ROOT");
+	if (strcpy_s(cfg_path, PATH_MAX, opae_path) != EOK) {
+		fprintf(stderr,
+				"Could not copy the opae_path successfully.\n");
+		return;
+	}
+
+	if (strcat_s(cfg_path, PATH_MAX, "/include/opae_ase.cfg") != EOK) {
```
Comment (tswhison) [inline libopae/init_ase.c]: We should follow the same config search path logic as opae-c here.

**Comment by lzhan55** [comment]
@tswhison it worked here. I can run hello_fpga with ASE.

**Comment by michael-adler** [comment]
@tswhison 's  question is good, though. If the relative order of loading libopae-c and libopae-c-ase is random, there was a 50% chance you'd get the order you needed during the test. We need to prove that either the order is unimportant or that it is deterministic. Unimportant would be better.

File: libopae/init_ase.c (hunk 1 above)
Comment (lzhan55) [reply libopae/init_ase.c]: @michael-adler I would put the opae_ase.cfg in the stand search locations for config file.

Comment (michael-adler) [reply libopae/init_ase.c]: I don't understand what that means.

A goal here should be that this keeps working with old releases, so we don't want to modify anything under OPAE_PLATFORM_ROOT.

On the other hand, your change is part of the OPAE SDK, so you are free to change whatever you need in the SDK sources or in the generated OPAE installation tree.

Comment (michael-adler) [reply libopae/init_ase.c]: I could be wrong, but from inspection it looks like the new code still isn't searching the installation directory. Aren't you going to create a default opae_ase.cfg and store it in <install>/etc/opae/opae_ase.cfg or (maybe better) under <install>/share/opae/ase? That would leave ASE completely functional with no user intervention.

If I'm reading the code correctly, it still requires a change by the user.

Comment (lzhan55) [reply libopae/init_ase.c]: Installed opae_ase.cfg to ~/.local/opae now so that ASE can work the same as before plugin changes.

Comment (michael-adler) [reply libopae/init_ase.c]: ~/.local assumes it is the ASE user who did the installation, which is not necessarily true. It won't work for a shared installation. Consider how the DCP test system works. OPAE is installed by a nightly script to some shared location. It is put on a PATH. The vLab is configured the same way.

I don't understand the resistance to putting a default copy of opae_ase.cfg in the place where ASE itself is stored. All our tools have to be able to find <install>/share/ase already. Isn't it the obvious place for a default opae_ase.cfg?

Comment (lzhan55) [reply libopae/init_ase.c]: It is already installed in the shared location share/opae/ase. I just need to put the shared location in the searching path. Is there an environment variable for the installation path? Probably OPAE_PLATFORM_ROOT can be used.

Comment (michael-adler) [reply libopae/init_ase.c]: There is no environment variable. The CMake build script knows where it is going to be installed and we use CMake variable substitution to construct sources that compile in the installation path. I don't know if there are C examples, but we do this for Python scripts like afu_sim_setup. Hence the reference to getASESrcPath() in the comments above.

The easiest way might be to create a header file that has the installation directory as a macro, add a CMake rule to do variable substitution on the header file, and put the build directory with the constructed header file on the C header search path.

File: CMakeLists.txt [hunk 7]
```diff
@@ -113,7 +113,7 @@ install(DIRECTORY common/include/opae
 install(DIRECTORY common/include/safe_string
   DESTINATION include
   COMPONENT safestrheaders)
-
+
```
Comment (tswhison) [inline CMakeLists.txt]: Remove unnecessary whitespace.

File: libopae/init_ase.c [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+	return NULL;
+}
+
+__attribute__((constructor)) STATIC void opae_ase_init(void)
+{
+	fpga_result res;
+	char *cfg_path = find_ase_cfg();
+
+	if (cfg_path == NULL)
+		OPAE_ERR("Could not find opae_ase.cfg file");
```
Comment (tswhison) [inline libopae/init_ase.c]: Should probably return here.  If not, then 113 will call fpgaInitialize with NULL, which isn't what you want; and 117 will attempt to free the NULL pointer.

Comment (lzhan55) [reply libopae/init_ase.c]: good catch. thanks.

File: cmake/config/config.h.in [hunk 9]
```diff
@@ -226,6 +226,8 @@
 #cmakedefine INTEL_FPGA_API_VERSION "@INTEL_FPGA_API_VERSION@"
 #cmakedefine INTEL_FPGA_API_HASH    "@INTEL_FPGA_API_HASH@"
 
+#define OPAE_ASE_CFG_SRC_PATH "@CMAKE_CURRENT_SOURCE_DIR@/libopae/plugins/ase/opae_ase.cfg"
```
Comment (michael-adler) [inline cmake/config/config.h.in]: I think what you wanted here was the equivalent of @ASE_INST_SHARE_DIR@, not the source. (See the Python example.) What you have is the source tree. What you want, I hope, is the installation tree.

Comment (lzhan55) [reply cmake/config/config.h.in]: @ASE_INST_SHARE_DIR@ equals to "/share/opae/ase". In Python, you found the current script directory which is concatenated with @ASE_INST_SHARE_DIR@. In FPGA sample execution, we cannot assume the executable's location since it could be anywhere. So it is not easy to get the installation location from executable's location.

Comment (michael-adler) [reply cmake/config/config.h.in]: No. We know exactly where ASE is installed. afu_sim_setup depends on it. I don't understand why this special ASE library can't also make the same inference. It's the only place where we can be confident there is a shared copy opae_ase.cfg that hasn't been installed as a special step by an admin.

Note also that this all has to work on a network of machines with a shared installation. Local disks, like /etc, don't cut it. Home directories require a change by the user.

Comment (lzhan55) [reply cmake/config/config.h.in]: @michael-adler

Comment (lzhan55) [reply cmake/config/config.h.in]: Added "@CMAKE_INSTALL_PREFIX@/share/opae/ase/opae_ase.cfg" to opae_ase.cfg search path.

File: libopae/init_ase.c [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+STATIC char *find_ase_cfg()
+{
+	int i = 0;
+	char *file_name = NULL;
+	char *opae_path = NULL;
+	char cfg_path[PATH_MAX];
+	char home_cfg[PATH_MAX] = {0};
+	char *home_cfg_ptr = &home_cfg[0];
+	// get the user's home directory
+	struct passwd *user_passwd = getpwuid(getuid());
```
Comment (tswhison) [inline libopae/init_ase.c]: Should probably check user_passwd for NULL (getpwuid can fail) and skip the checks starting at 101 if NULL.

Comment (lzhan55) [reply libopae/init_ase.c]: OK
